runtime/memprof.c

```diff
@@ -206,1126 +1258,921 @@ void caml_memprof_update_suspended(bool s) {
+} callstack_stash_s, *callstack_stash_t;
+
+/* If the callstack buffer is large and we've only used a small part
+ * of it, free it. This saves us from C heap bloat due to unbounded
+ * lifetime of the callstack buffers (as callstacks may sometimes be
+ * huge). */
+static void shrink_callstack_buffer(memprof_domain_t domain, size_t frames)
 {
-  value res;
-  intnat callstack_len =
-    caml_collect_current_callstack(&callstack_buffer, &callstack_buffer_len,
-                                   callstack_size, -1);
-  if (callstack_len == 0)
-    return Atom(0);
-  res = caml_alloc_shr_no_track_noexc(callstack_len, 0);
-  if (res == 0)
-    return Atom(0);
-  memcpy(Op_val(res), callstack_buffer, sizeof(value) * callstack_len);
-  if (callstack_buffer_len > 256 && callstack_buffer_len > callstack_len * 8) {
-    caml_stat_free(callstack_buffer);
-    callstack_buffer = NULL;
-    callstack_buffer_len = 0;
+  if (domain->callstack_buffer_len > 256 &&
+      domain->callstack_buffer_len > frames * 8) {
+    caml_stat_free(domain->callstack_buffer);
+    domain->callstack_buffer = NULL;
+    domain->callstack_buffer_len = 0;
   }
-  return res;
 }
 
-/* In this version, we are allowed to call the GC, so we use
-   [caml_alloc], which is more efficient since it uses the minor
-   heap.
-   Should be called with [local->suspended == 1] */
-static value capture_callstack(int alloc_idx)
+/* Capture the call stack when sampling an allocation from the
+ * runtime. We don't have to account for combined allocations
+ * (Comballocs) but we can't allocate the resulting stack on the Caml
+ * heap, because the heap may be in an invalid state so we can't cause
+ * a GC. Therefore, we capture the callstack onto the C heap, and will
+ * copy it onto the Caml heap later, when we're ready to call the
+ * allocation callback. */
+static value capture_callstack_no_GC(memprof_domain_t domain)
 {
-  value res;
-  intnat callstack_len =
-    caml_collect_current_callstack(&callstack_buffer, &callstack_buffer_len,
-                                   callstack_size, alloc_idx);
-  CAMLassert(local->suspended);
-  res = caml_alloc(callstack_len, 0);
-  memcpy(Op_val(res), callstack_buffer, sizeof(value) * callstack_len);
-  if (callstack_buffer_len > 256 && callstack_buffer_len > callstack_len * 8) {
-    caml_stat_free(callstack_buffer);
-    callstack_buffer = NULL;
-    callstack_buffer_len = 0;
+  value res = Atom(0); /* empty array. */
+  size_t frames =
+    caml_get_callstack(Callstack_size(domain->entries.config),
+                       &domain->callstack_buffer,
+                       &domain->callstack_buffer_len, -1);
+  if (frames) {
+    callstack_stash_t stash = caml_stat_alloc_noexc(sizeof(callstack_stash_s)
+                                                    + frames * sizeof(value));
+    if (stash) {
+      stash->frames = frames;
+      for (size_t i = 0; i < frames; ++i) {
+        stash->stack[i] = Val_backtrace_slot(domain->callstack_buffer[i]);
```

Comment:
Since this doesn't really need to be encoded as a value (it's in off-heap data), it might be simpler to leave `stash->stack` as a `backtrace_slot*` and just `memcpy` here, leaving the `Val_backtrace_slot` encoding for when it arrives in the heap in `unstash_callstack`.